ttnctl/cmd/devices_export.go

```diff
@@ -0,0 +1,171 @@
+// Copyright © 2019 The Things Network
+// Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/TheThingsNetwork/api"
+	"github.com/TheThingsNetwork/api/handler"
+	"github.com/TheThingsNetwork/go-utils/log"
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+func exportV3Device(dev *handler.Device, flags *pflag.FlagSet) *v3Device {
+	v3Dev := &v3Device{}
+	v3Dev.IDs.ApplicationIDs.ApplicationID = dev.AppID
+	v3Dev.IDs.DeviceID = dev.DevID
+
+	if dev.Latitude != 0 || dev.Longitude != 0 {
+		v3Dev.Locations = map[string]v3DeviceLocation{
+			"user": {
+				Latitude:  float64(dev.Latitude),
+				Longitude: float64(dev.Longitude),
+				Altitude:  dev.Altitude,
+				Source:    "SOURCE_REGISTRY",
+			},
+		}
+	}
+
+	v3Dev.Name = dev.DevID
+	v3Dev.Description = dev.Description
+	v3Dev.Attributes = dev.Attributes
+
+	lorawanDevice := dev.GetLoRaWANDevice()
+
+	v3Dev.IDs.JoinEUI = lorawanDevice.GetAppEUI().String()
+	v3Dev.IDs.DevEUI = lorawanDevice.GetDevEUI().String()
+	if lorawanDevice.AppKey.String() != "" {
+		v3Dev.RootKeys = &v3DeviceRootKeys{
+			AppKey: v3DeviceKey{Key: lorawanDevice.AppKey.String()},
+		}
+	}
+	if devAddr := lorawanDevice.DevAddr; !devAddr.IsEmpty() {
+		v3Dev.DevAddr = devAddr.String()
+		v3Dev.Session = &v3DeviceSession{}
+		v3Dev.Session.DevAddr = lorawanDevice.DevAddr.String()
+		if lorawanDevice.AppSKey.String() != "" {
+			v3Dev.Session.Keys.AppSKey = v3DeviceKey{Key: lorawanDevice.AppSKey.String()}
+		}
+		if lorawanDevice.NwkSKey.String() != "" {
+			v3Dev.Session.Keys.FNwkSIntKey = v3DeviceKey{Key: lorawanDevice.NwkSKey.String()}
+		}
+		v3Dev.Session.LastFCntUp = lorawanDevice.FCntUp
+		v3Dev.Session.LastNFCntDown = lorawanDevice.FCntDown
+	}
+
+	v3Dev.MACSettings.Supports32BitFCnt = lorawanDevice.Uses32BitFCnt
+	v3Dev.MACSettings.ResetsFCnt = lorawanDevice.DisableFCntCheck
+
+	v3Dev.MACVersion = "MAC_V1_0_2"
+	v3Dev.PHYVersion = "PHY_V1_0_2_REV_B"
+
+	if lorawanDevice.AppKey.String() == "" {
+		v3Dev.SupportsJoin = false
+		v3Dev.MACSettings.Rx1Delay = "RX_DELAY_1"
+	} else {
+		v3Dev.SupportsJoin = true
+		if !lorawanDevice.AppSKey.IsEmpty() && !lorawanDevice.NwkSKey.IsEmpty() && !lorawanDevice.DevAddr.IsEmpty() {
+			v3Dev.MACState.CurrentParameters.Rx1Delay = "RX_DELAY_1"
+		}
+	}
+
+	frequencyPlanFlag, _ := flags.GetString("frequency-plan-id")
+	frequencyPlan, err := getOption(frequencyPlans, frequencyPlanFlag, "")
+	if err != nil {
+		ctx.WithError(err).WithFields(log.Fields{
+			"frequency_plan_id":            frequencyPlanFlag,
+			"available_frequency_plan_ids": frequencyPlans,
+		}).Fatal("Invalid --frequency-plan-id argument")
+	}
+	v3Dev.FrequencyPlanID = frequencyPlan
+
+	v3Dev.SupportsClassB, _ = flags.GetBool("class-b")
+	v3Dev.SupportsClassC, _ = flags.GetBool("class-c")
```

Comment:
Can remove these, because they are always false. V2 never had support for class B and C.

Reply:
Done.